### ISSUE TITLE
Non-global path exclusion

### ISSUE THREAD

**Issue by paul1149** [cboxdoerfer/fsearch#522]
I have a drive I use for storage. One directory hierarchy on it is "backup", which contains several incremental /home backups. 

Consequently, when I do a search of both /home and the storage drive, often many hits will come from the backup directory, visually spamming and complicating the search results.

I would like to create an exclusion for this directory to avoid this. But I also would like to be able to search it when necessary.

Currently (FS 0.2.3), I can create the exclusion, but it is global. I can toggle it on/off, but doing so means going into the Preferences, and then waiting for the DB to update itself, and then trying to remember next time what state I left it in..

The way to solve this, AISI, is either to allow REGEX in the DB INclusions, so the backup directory could be excluded that way using a negation, or to list exclusions in a toolbar dropdown box with checkboxes to toggle them, without necessitating a DB update.

Am I making sense? And is there a way to do this currently?

Thanks much.

**Comment by cboxdoerfer** [comment]
Hi, currently there's no way to simply toggle exclusions on and off. But you can create custom filters in Preferences -> Search, which you then can easily switch between in the toolbar. There are different approaches to that, one could be:

This filter allows you search everywhere but the backup folder:

   Name: `NoBackup` (you can name this how you want)
   Search in path: `x`
   Match case: `x`
   Query: `NOT regex:^/path/to/backup/folder`


and the second filter allows you to only search in the backup folder:

   Name: `Backup`
   Search in path: `x`
   Match case: `x`
   Query: `regex:^/path/to/backup/folder`

I hope that solves the issue

**Comment by paul1149** [comment]
Thanks.

I've got this search definition:

Search in path [x]
Query: NOT regex:^/home/paul/media

And yet when I run the search it is picking up hits in the ~/media hierarchy.

I've also tried, with Regex enabled:

regex:^/home\/paul/(?!media).*
regex:^\/home\\/paul\/(?!media).*
regex:"^/home\/paul/(?!media).*"

But I'm getting nowhere.

Then under the Wiki here I saw that Regex are on the todo list? Maybe that's outdated.

Finally, a different problem. When I have a list of search hits, if I cilk on the Path column to sort them, the sort is faulty. I'll get some /home/paul... hits, then some /media/.... hits, and then more /home/paul... hits. 

I'll upload a shot of that. Sorry for so many problems.

![fs sort](https://github.com/cboxdoerfer/fsearch/assets/13952246/7f51a5c2-5aff-4996-bd9a-6b4c0dd77013)

**Comment by cboxdoerfer** [comment]
> I've got this search definition:
>
> Search in path [x]
> Query: NOT regex:^/home/paul/media
> 
> And yet when I run the search it is picking up hits in the ~/media hierarchy.

Can you post a screenshot of that filter and maybe a screenshot of results that are not supposed to show up with that filter?

> Then under the Wiki here I saw that Regex are on the todo list? Maybe that's outdated.

The TODO only refers to the documentation. The regex feature is fully implemented and working.

> Finally, a different problem. When I have a list of search hits, if I cilk on the Path column to sort them, the sort is faulty. I'll get some /home/paul... hits, then some /media/.... hits, and then more /home/paul... hits.

The order is as intended, because folders and files are always listed separately. So you first get all folders sorted by path and then all files sorted path.

> Sorry for so many problems.

No problem.

**Comment by paul1149** [comment]
Ok, thanks for explaining the wiki entry and the folders sorting. I hadn't thought about folders first at all, since the path is so long it truncated the last element in the paths.  Now to the filter.

Here's my Home only filter:

![fs w regex](https://github.com/cboxdoerfer/fsearch/assets/13952246/3f14e2fd-c117-41cf-b5b4-216df473a7f5)

Here are the results:

![fs w regex -empty](https://github.com/cboxdoerfer/fsearch/assets/13952246/8ece112d-0414-44a6-b6be-94c90e208288)

If I toggle regex off, it's still empty:

![fs w regex toggled off - still empty](https://github.com/cboxdoerfer/fsearch/assets/13952246/ef6b9b2c-4117-443a-81d3-6212c4423857)

But if I go back to the filter and turn off regex there:

![fs regex turned off at filter](https://github.com/cboxdoerfer/fsearch/assets/13952246/ae95f6cd-4dc0-4607-892a-d29a87dc1f33)

Then I get hits, but the undesired / media... is included

![fs no regex -wrong path](https://github.com/cboxdoerfer/fsearch/assets/13952246/65808474-5338-4a82-bc54-ffa6bd8b062b)

And here's a shot sorted to show that I should be getting hits from the home directory:

![fs hits in home directory](https://github.com/cboxdoerfer/fsearch/assets/13952246/924a55fe-cb4c-4968-936f-519a7bcaa03c)

Thanks again.

**Comment by cboxdoerfer** [comment]
> Here's my Home only filter:

The regular expression checkbox needs to be unchecked, otherwise the whole query will be interpreted as a regular expression and `NOT` is not a valid operator in regex.

> But if I go back to the filter and turn off regex there:
> 
> ...
> 
> Then I get hits, but the undesired / media... is included

That's to be expected. If you want to exclude the `/media` directory, the query entry should look like this:

`NOT regex:^/media/`

You should also check the `Match case` checkbox, because this improves the performance of the filter.

So to summarize:

Name: HomeOnly
Search in path: Yes
Match case: Yes
Query: `NOT regex:^/media/`

Alternatively, if you really only want to search in your home folder, you can also use the following query instead:

Query: `regex:^/home/paul/`

**Comment by paul1149** [comment]
OK, I see one of my big problems. At one time the secondary drives were mounted under my home directory. They are now under /media, but I had forgotten that. Let me process that and get back to you.

Thanks much.

**Comment by paul1149** [comment]
Ok, I think I've got it pretty much figured out, and it is working well. I had trouble getting a regex with an exclusion like `(?!/media/NVMe930/backups` to work, but in the end I think coupling several expression with logical operators is more flexible:
![nvme filter](https://github.com/cboxdoerfer/fsearch/assets/13952246/32028992-a164-4891-8f74-62aea4344935)

It's all working perfectly now. Thanks much for your help, and I hope you have a great holidays.
